pkg/foundation/grpcutil/websocket.go

```diff
@@ -114,12 +113,20 @@ func (p *webSocketProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 // and stop writing any data to the WebSocket connection. This will
 // automatically halt all the "pipeline nodes" after the underlying response.
 func (p *webSocketProxy) proxy(w http.ResponseWriter, r *http.Request) {
-	ctx, cancelFn := context.WithCancel(r.Context())
-	defer cancelFn()
+	ctx, cancelCtx := context.WithCancel(r.Context())
+	defer cancelCtx()
 	r = r.WithContext(ctx)
 	go func() {
-		<-p.done
-		cancelFn()
+		// we're using ChanOut.Recv() so that the goroutine doesn't wait on
+		// the proxy to be done even when the request is done
+		_, ok, err := cchan.ChanOut[struct{}](p.done).Recv(ctx)
+		if err != nil {
+			p.logger.Warn(ctx).Msgf("request context returned an error: %v", err)
+		}
+		if ok {
+			p.logger.Info(ctx).Msg("shutting down")
+		}
```

Comment:
I think you have a point about the "shutting down". For the warning above, it was because there's a possibility that `ctx.Err()` returns a `DeadlineExceeded`, which I believe is a bit unexpected. So I went the route of logging a bit more, to catch that as well.